docs/design/operator.md

```diff
@@ -0,0 +1,414 @@
+# Ceph CSI Operator Design Document
+
+## Introduction
+
+This document outlines the design of the Ceph-CSI Operator, which aims to
+provide easy management of Ceph-CSI drivers (CephFS, RBD, and NFS) in
+Kubernetes based environments. The operator automates the deployment,
+configuration, and management of these drivers using Custom Resource
+Definitions (CRDs).
+
+## System Overview
+
+The Ceph-CSI Operator is a Kubernetes operator that simplifies the management
+of Ceph CSI drivers. It provides a set of CRDs to manage the configurations of
+CephFS, RBD, and NFS drivers, as well as connection details to Ceph clusters.
+The operator ensures that the desired state specified in the CRDs is maintained
+in the cluster.
+
+## Assumptions and Dependencies
+
+Kubernetes cluster is available and operational. Ceph cluster is set up and
+accessible from the Kubernetes cluster. Users are familiar with Kubernetes and
+Ceph concepts. CRD and operator versions are compatible with the Kubernetes
+version in use.
+
+## General Constraints
+
+Resource limitations of the Kubernetes cluster (CPU, memory). Network latency
+and connectivity between the Kubernetes cluster and Ceph cluster. Watching for
+changes on the ceph cluster and updating internal details.  Security policies
+and RBAC configurations in the Kubernetes cluster.
+
+## Goals and Guidelines
+
+Automate the installation and management of Ceph-CSI drivers. Provide a
+consistent and flexible configuration mechanism using CRDs. Enable
+namespace-scoped configurations for different Ceph-CSI drivers. Ensure high
+availability and scalability of the operator. When running operator in a
+namespace provide an option to watch for the CR's in the  namespace where its
+running for ceph-CSI management. User will have an option to deploy multiple
+ceph-csi-operator that are configured to watch for the namespaces where they
+are deployed.
+
+## Development Methods
+
+Use operator framework which is higly used in operator developed and also by
+following Kubernetes best practices for operator development.
```

Comment:
I want to revisit this, I am now in favor of using `kubebuilder`

Reply:
Lets discuss about pros and cons, this was based on the last few discussions.